### PR TITLE
ENG-8434 compose text watcher

### PR DESCRIPTION
**Pull request by kishisaka-nid**
Limitation on paste detection, text must be at the end of existing text in a field or into a blank field. This is an existing limitation in the current SDK. Each field must be individually monitored. 

calls:
```
NeuroID.getInstance()?.compose?.trackElement(elementState = firstName, elementName = "firstName", pageName = "PersonalInfo")
NeuroID.getInstance()?.compose?.trackElement(elementState = lastName, elementName = "lastName", pageName = "PersonalInfo")
```

events: 
```
2024-08-06 11:15:42.601 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: TOUCH_MOVE - main_view - xy=[NIDTouchModel(tid=0.0, x=348.0, y=810.0)] tg={tgs=main_view, sender=, etn=} tgs=main_view ec=null syn=null
2024-08-06 11:15:42.991 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~1, tg={attr=[{n=v, v=S~C~~1}, {n=hash, v=8009acd3}], etn=INPUT, et=text}
2024-08-06 11:15:43.104 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~2, tg={attr=[{n=v, v=S~C~~2}, {n=hash, v=13600eaf}], etn=INPUT, et=text}
2024-08-06 11:15:43.180 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~3, tg={attr=[{n=v, v=S~C~~3}, {n=hash, v=3675f147}], etn=INPUT, et=text}
2024-08-06 11:15:43.298 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~4, tg={attr=[{n=v, v=S~C~~4}, {n=hash, v=23e37c0b}], etn=INPUT, et=text}
2024-08-06 11:15:43.367 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~5, tg={attr=[{n=v, v=S~C~~5}, {n=hash, v=29354a1e}], etn=INPUT, et=text}
2024-08-06 11:15:43.433 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~6, tg={attr=[{n=v, v=S~C~~6}, {n=hash, v=02a2372d}], etn=INPUT, et=text}
2024-08-06 11:15:43.507 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~8, tg={attr=[{n=v, v=S~C~~8}, {n=hash, v=af84954c}], etn=INPUT, et=text}
2024-08-06 11:15:43.957 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: TOUCH_MOVE - main_view - xy=[NIDTouchModel(tid=0.0, x=196.0, y=870.0)] tg={tgs=main_view, sender=, etn=} tgs=main_view ec=null syn=null
2024-08-06 11:15:43.974 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: TOUCH_MOVE - main_view - xy=[NIDTouchModel(tid=0.0, x=196.0, y=870.0)] tg={tgs=main_view, sender=, etn=} tgs=main_view ec=null syn=null
.
.
.	  
2024-08-06 11:15:49.009 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: PASTE - PersonalInfo:firstName - 
2024-08-06 11:15:49.014 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:firstName - v=S~C~~20, tg={attr=[{n=v, v=S~C~~20}, {n=hash, v=5eccd1ec}], etn=INPUT, et=text}
2024-08-06 11:15:49.662 31161-31339 NeuroID Debug Payload:  com...uroid_android_compose_sandbox  D  Payload Summary
.
.
.
2024-08-06 11:21:12.529 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~9, tg={attr=[{n=v, v=S~C~~9}, {n=hash, v=5699e493}], etn=INPUT, et=text}
2024-08-06 11:21:12.588 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~8, tg={attr=[{n=v, v=S~C~~8}, {n=hash, v=af84954c}], etn=INPUT, et=text}
2024-08-06 11:21:12.630 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~7, tg={attr=[{n=v, v=S~C~~7}, {n=hash, v=0f300f29}], etn=INPUT, et=text}
2024-08-06 11:21:12.685 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~6, tg={attr=[{n=v, v=S~C~~6}, {n=hash, v=02a2372d}], etn=INPUT, et=text}
2024-08-06 11:21:12.726 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~5, tg={attr=[{n=v, v=S~C~~5}, {n=hash, v=29354a1e}], etn=INPUT, et=text}
2024-08-06 11:21:12.784 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~4, tg={attr=[{n=v, v=S~C~~4}, {n=hash, v=23e37c0b}], etn=INPUT, et=text}
2024-08-06 11:21:12.832 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~3, tg={attr=[{n=v, v=S~C~~3}, {n=hash, v=3675f147}], etn=INPUT, et=text}
2024-08-06 11:21:12.884 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~2, tg={attr=[{n=v, v=S~C~~2}, {n=hash, v=13600eaf}], etn=INPUT, et=text}
2024-08-06 11:21:12.930 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~1, tg={attr=[{n=v, v=S~C~~1}, {n=hash, v=8009acd3}], etn=INPUT, et=text}
2024-08-06 11:21:12.986 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~0, tg={attr=[{n=v, v=S~C~~0}, {n=hash, v=3ef37eb6}], etn=INPUT, et=text}
2024-08-06 11:21:14.953 31161-31473 NeuroID Debug Payload:  com...uroid_android_compose_sandbox  D  Payload Summary
.
.
.
2024-08-06 11:21:23.890 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: PASTE - PersonalInfo:lastName - 
2024-08-06 11:21:23.893 31161-31161 NeuroID Debug Event     com...uroid_android_compose_sandbox  D  EVENT: INPUT - PersonalInfo:lastName - v=S~C~~10, tg={attr=[{n=v, v=S~C~~10}, {n=hash, v=e659280c}], etn=INPUT, et=text}
2024-08-06 11:21:24.958 31161-31473 NeuroID Debug Payload:  com...uroid_android_compose_sandbox  D  Payload Summary

```

### CODE REVIEW

File: NeuroID/src/main/java/com/neuroid/tracker/compose/JetpackCompose.kt [hunk 1]
```diff
@@ -82,4 +85,58 @@ class JetpackComposeImpl(
...
+                        hashMapOf(
+                            "attr" to getAttrJson(elementState),
+                            "etn" to INPUT,
+                            "et" to "text",
+                        ),
+                    tgs = "$pageName:$elementName",
+                    v = "S~C~~${elementState.length}",
+                    hv = elementState.getSHA256withSalt().take(8),
+                )
+                previousText = elementState
```
Comment (priya-xavier) [inline NeuroID/src/main/java/com/neuroid/tracker/compose/JetpackCompose.kt:138]: We could also add the `attrs` field to show that its from a compose element like the button and page events? attrs = listOf(
                sdkMap
            )
where sdkMap =mapOf(
        "jetpackCompose" to true,
    )

Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/compose/JetpackCompose.kt:142]: added. thanks!

File: NeuroID/src/main/java/com/neuroid/tracker/utils/NIDComposeTextWatcherUtils.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        // no paste event if we get here
+        return false
+    }
+
+    fun getTextChange(
+        previousText: String,
+        currentText: String,
+    ): TextChange {
+        // equivalent of the TextWatcher, we need the count and changedText
+        val start = currentText.commonPrefixWith(previousText).length
```
Comment (cometman) [inline NeuroID/src/main/java/com/neuroid/tracker/utils/NIDComposeTextWatcherUtils.kt:33]: What happens when the current text change is less than the starting length. 

For example, when I run `getTextChange("zz", "")`

The result is: `TextChange(start = 0, count = 0, changedText = "")` Which doesnt seem. quite right (or I could me misunderstanding it)

**Review by cometman — CHANGES_REQUESTED** [review]

File: NeuroID/src/main/java/com/neuroid/tracker/utils/NIDComposeTextWatcherUtils.kt (hunk 2 above)
Comment (kishisaka-nid) [reply NeuroID/src/main/java/com/neuroid/tracker/utils/NIDComposeTextWatcherUtils.kt:33]: You delete two characters so the count of the left over characters is 0, the start is at 0 index and the changedText  is basically the deletions which is "". That is what I am seeing. I'll add a test case for deletions. For our case this is enough info to calculate pastes.

**Review by cometman — APPROVED** [review]

**Review by priya-xavier — APPROVED** [review]

**Pull request merged**